### PR TITLE
Client: Rework Client Map

### PR DESCRIPTION
**Pull request by psiinon**
## Overview
Reworked (rewrote?;) the Client Map to handle parameters correctly and various edge cases, inc fragments on URLs with parameters.

## Checklist
- [ ] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

**Comment by psiinon** [comment]
Will fix the conflicts ..

**Comment by psiinon** [comment]
Fixed

**Review by kingthorin — COMMENTED** [review]
Looks good overall.

File: addOns/client/src/main/java/org/zaproxy/addon/client/ClientNode.java [hunk 1]
```diff
@@ -58,6 +78,46 @@ public String getSite() {
...
+
+    /**
+     * A diagnostic method which returns a text representation of the node.
+     *
+     * @param summary the list to use, if null a new list will be created
+     * @param indent
+     * @param recurse
+     * @return
+     */
+    public List<String> getNodeSummary(List<String> summary, int indent, boolean recurse) {
```
[the remark this answers is not in the document]
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ClientNode.java:100]: Its a diagnostic method which does no harm 😁 
I've been using it from a standalone script and I dont want to have to use reflection to be able to access it.
Why would we want to make it hard to access?

[the remark this answers is not in the document]
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ClientNode.java:100]: happy to improve the javadocs though..

[the remark this answers is not in the document]
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ClientNode.java]: updated

[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ClientNode.java]: Thanks

**Review by kingthorin — APPROVED** [review]

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!
